lisp/tree-sitter-highlight.el

```diff
@@ -0,0 +1,334 @@
+           x)))))
+
+(defun tree-sitter-highlight--get-matches (start end)
+  (ts-set-point-range tree-sitter-highlight--query-cursor
+    (ts--point-from-position start)
+    (ts--point-from-position end))
+  (ts--query-cursor-matches
+    tree-sitter-highlight--query-cursor
+    tree-sitter-highlight--query
+    (ts-root-node tree-sitter-tree)
+    nil
+    ;; t
+    #'ts-node-text)
+  )
+
+(defun tree-sitter-highlight--jit (old-tree)
+  (when old-tree
+    (let ((changes (ts-changed-ranges old-tree tree-sitter-tree))
+           (wstart (window-start))
+           (wend (window-end)))
+      ;; TODO: Remember what we've highlighted, similar to how font-lock does it.
+      ;;       Already highlighted regions shouldn't be re-highlighted.
+
+      ;; Find changes that are within the current window
+      (mapc #'(lambda (range)
+                (let ((start (aref range 0))
+                       (end (aref range 1)))
+                  ;; TODO: Improve this
+                  (tree-sitter-highlight--highlight (max wstart start) (min wend end))))
+        changes))))
+
+(defun tree-sitter-highlight--highlight (start end)
+  (ts--save-context
+    (with-silent-modifications
+      (remove-text-properties start
+        end
+        '(face nil))
+      (let ((matches (tree-sitter-highlight--get-matches start end)))
+        (seq-do #'(lambda (match)
+                    (seq-do #'tree-sitter-highlight--apply (cdr match)))
+          matches)))))
+;; (seq-do #'tree-sitter-highlight--apply
+
+;; (ts-highlight (ts-root-node tree-sitter-tree)
+;;               tree-sitter-highlight--query-cursor
+;;               tree-sitter-highlight--query
+;;               tree-sitter-highlight--injections-query
+;;               #'ts-node-text
+;;               (lambda (start end)
+;;                 (remove-text-properties (ts-byte-to-position start)
+;;                                         (ts-byte-to-position end)
+;;                                         '(face nil)))
+;;               (lambda (language)
+;;                 (car (tree-sitter-highlight--get-injection (intern language))))
+;;               (lambda (language)
+;;                 (cadr (tree-sitter-highlight--get-injection (intern language))))
+;;               (ts-byte-from-position start)
+;;               (ts-byte-from-position end)))))
+
+(defun tree-sitter-highlight--highlight-window (_window start)
+  (tree-sitter-highlight--highlight start (window-end nil t)))
+
+(defun tree-sitter-highlight--enable ()
+  "Enable `tree-sitter-highlight' in this buffer."
+  (run-hooks  'tree-sitter-highlight-setup-functions)
+  (setq tree-sitter-highlight--face-hash
+    (let ((table (make-hash-table :test 'equal)))
+      (mapc (lambda (x)
+              (pcase-let ((`(,key . ,value) x))
+                (puthash key value table)))
+        tree-sitter-highlight-default-faces)
+      table))
+  (let ((x (tree-sitter-highlight--make-queries (alist-get major-mode
```

Comment:
@ubolonton not related to this PR but IMO tree-sitter should not be on top of the existing major modes but it should be a replacement for them. So instead of having major-mode -> language mapping IMO it should be "file-regex" -> language.

Reply:
When there are tree-sitter-based major modes, the mapping mechanism would simply be `auto-mode-alist`.

However, that's likely too far away in the future. For adoption, I think it's better to piggyback on major modes.

Reply:
I think we can just use `tree-sitter-language` here though, and leave the mapping for `tree-sitter-mode`.

Reply:
Also, maybe we can define a `tree-sitter-highlight-queries` var, as the replacement of `font-lock-keywords`.

Language-specific minor modes (e.g. `tree-sitter-rust`) would then be responsible for loading/populating the queries however they see fit.

Reply:
Just want to raise one more point: major-mode -> language mapping is not N -> 1 but it is N -> N. We had that issue in lsp-mode - for example, web-mode could be html, javascript, typescript, and so on. 

Also, why do you think that it is too far in the future? I mean except highlighting and eventually indentation what are the other things that will be missing to have tree-sitter centric major modes?

Reply:
> Just want to raise one more point: major-mode -> language mapping is not N -> 1 but it is N -> N. We had that issue in lsp-mode - for example, web-mode could be html, javascript, typescript, and so on.

Yeah, that's definitely an issue. I haven't thought about multi-language files in-depth.

> Also, why do you think that it is too far in the future? I mean except highlighting and eventually indentation what are the other things that will be missing to have tree-sitter centric major modes?

I think the main technical barrier is that many Emacs distributions have dynamic module support turned off. Other than that, I was mainly thinking about adoption/mindshare.